[Expo Router V3] Fix route pushing

kennethstarkrl opened the pull request:
# Why

Was still having issues with pushing even after v3 latest changes, where router.push() was still performing as router.pushOrPop(). 

# How

When users directly use router.push() the route should be pushed - it shouldn't go back down through all the states and pop the last state, it should just push to the stack. 

# Test Plan

Verify when router.push() is used, that the new route is always pushed to the stack.

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

kennethstarkrl committed:
Update routing.js

Fixes issues with pushing the state. If we specify a push, we don't want to continue down the tree.

kennethstarkrl committed:
Merge pull request #1 from kennethstarkrl/kennethstarkrl-expo-router-fix-push

Update routing.js

kennethstarkrl committed:
Update routing.ts src

kennethstarkrl committed:
Update CHANGELOG.md

kennethstarkrl committed:
Update navigation.test.ios.tsx

kennethstarkrl committed:
Merge branch 'main' into main

marklawlor commented on packages/expo-router/src/global-state/routing.ts:
```diff
@@ -145,7 +145,8 @@ function getNavigateAction(state: ResultState, parentState: NavigationState, typ
   // If there is nested state and the routes are equal, we should keep going down the tree
-  if (route.state && routesAreEqual && currentRoute.state) {
+  // unless the type is push, in which case we should just push the state
+  if (route.state && routesAreEqual && currentRoute.state && type !== 'PUSH') {
```
`type !== 'PUSH'` seems wrong as we have this guard just below

https://github.com/expo/expo/blob/b5ad7c606700c0bff217ee036bf032c93adcc1c6/packages/expo-router/src/global-state/routing.ts#L158

kennethstarkrl replied:
could be that it's because I'm using a drawer and not tabs?

kennethstarkrl replied:
getting rid of this seems to fix my issue as well. I'm not sure why it needs to change the type if on the function call it defaults to NAVIGATE anyways.
https://github.com/expo/expo/blob/b5ad7c606700c0bff217ee036bf032c93adcc1c6/packages/expo-router/src/global-state/routing.ts#L156-L157

kennethstarkrl replied:
never mind, that's not entirely true.

kennethstarkrl replied:
It has something to do with routesAreEqual. It's comparing matching states when the route in the new state doesn't match. The only thing changing is the params. In the screen shot the sort param changed but the route doesn't push.
![image](https://github.com/expo/expo/assets/53270830/0b333ff6-31e6-4466-a950-2f789bc9b8e8)

kennethstarkrl replied:
stepping away from this before I get homicidal

kennethstarkrl replied:
Ok, so I think I found the pain points

First is routesAreEqual
https://github.com/expo/expo/blob/b5ad7c606700c0bff217ee036bf032c93adcc1c6/packages/expo-router/src/global-state/routing.ts#L145

routesAreEqual is returning true even though they are not the same. This is because objects are compared by reference, not by value. When comparing objects using == or ===, JavaScript checks to see if the objects reference the same location in memory, not if their contents are the same. Which in this case matters because the params have changed.

Second is currentRoute
https://github.com/expo/expo/blob/b5ad7c606700c0bff217ee036bf032c93adcc1c6/packages/expo-router/src/global-state/routing.ts#L144

This returns the first route in the list that matches the name, so when doing the comparison, it's comparing the first route rather than the last route (current route)

I think I'm close to a working solution with the functionality intended which is not to push the route if it is exactly the same.

marklawlor commented on packages/expo-router/src/__tests__/navigation.test.ios.tsx:
```diff
@@ -810,6 +810,68 @@ it('can push the same route multiple times', () => {
 });
 
+it('can continue to push with same path but different params', async () => }
+  renderRouter({
```
This test is pass on the current `main` branch without your changes. I'm not sure it correctly describes the behaviour your fixing.

kennethstarkrl replied:
I think I had to change it to a drawer. Commit: https://github.com/expo/expo/pull/26597/commits/7f79de417f650675bd2752ae44d92a2eb0671857

kennethstarkrl committed:
Update navigation.test.ios.tsx

kennethstarkrl committed:
Update routing.ts

- Fixed currentRoute to grab the last route from the parentRoute, previously it was comparing against the first parentRoute.
- Added deep object comparison for true equality. Now goes in and deeply compares the state objects and params to see if they are the same and skip object keys we don't need for the comparison.

kennethstarkrl committed:
Update routing.js

kennethstarkrl commented:
@marklawlor I believe the latest commit https://github.com/expo/expo/pull/26597/commits/f2a0c2f0628f714a34b68312c0361f0faa985ded should fix the issues. I'm not quite sure how to write a test case exactly for this issue as it depends on when the route is updaing and when useEffect fires to grab new data.

It now will push the state if route params change too. It still retains checking if the routes are exactly equal and will not push if so.

kennethstarkrl commented:
The issue still occurs with path queries ie /path/to/route?**query=test**. Route still doesn't push when the url query param changes. Has something to do with the parent state nesting params, where the new param shows in the nested, but not in the main object..

For example, sometimes it will appear like this
```js
currentRoute.params = {
  path:'a',
  to:'b',
  route:'c',
  params: {
    path:'a',
    to:'b',
    route:'c',
    query:'test'
  }
}
```
